types/coin.go

```diff
@@ -158,6 +158,26 @@ func (coins Coins) IsValid() bool {
 	}
 }
 
+// AddCoinByDenom looks for a matching coin in coins that has the same denom as
+// other. Upon matching, the other coin is added to the matching coin. If the
+// coins are empty, the coin is simply added to the empty set.
+func (coins Coins) AddCoinByDenom(other Coin) Coins {
+	res := copyCoins(coins)
+
+	if len(res) == 0 {
+		return Coins{other}
+	}
+
+	for i, coin := range res {
```

Comment:
We could binary search here.

Maybe premature optimization though...

Reply:
Agreed. I do think it's a premature optimization though. I have added a TODO to both methods and will create an issue for a `Coin` binary search on `Coins` 👍